tests/isa-test.cc

```diff
@@ -0,0 +1,774 @@
+
+	template<typename T>
+	void test_restrict_capability(Capability<T> capability, PermissionSet mask)
+	{
+		PermissionSet requestedPermissions(mask);
+		requestedPermissions &= capability.permissions();
+		PermissionSet expectedPermissions =
+		  permissions_to_representable(requestedPermissions);
+		capability.permissions() &= mask;
+		TEST(capability.permissions() == expectedPermissions,
+		     "permissions {} did not match expected {}",
+		     capability.permissions(),
+		     expectedPermissions);
+	}
+
+	void test_and_perms()
+	{
+		int              myLocal = 0;
+		const Capability CapabilityToLocal{&myLocal};
+		const Capability CapabilityToGlobal{myGlobalIntPointer};
+		const Capability CapabilityToFunction{
+		  reinterpret_cast<const void *>(&test_and_perms)};
+		debug_log("Checking and perms results...");
+		// TODO we could do with a sealing capability too, but then this would
+		// need to become a privileged compartment.
+		// Test all possible masks on our 'root' capabilities. For now we only
+		// test permissions on data / executable capabilities. Conveniently
+		// these occupy the lower 9 bits of the permissions field.
+		for (uint32_t p = 0; p <= 0x1ff; p++)
+		{
+			auto permissionMask = PermissionSet::from_raw(p);
+			test_restrict_capability(CapabilityToLocal, permissionMask);
+			test_restrict_capability(CapabilityToGlobal, permissionMask);
+			test_restrict_capability(CapabilityToFunction, permissionMask);
+		}
+	}
+
+	[[cheri::interrupt_state(disabled)]] void *get_interrupts_disabled_sentry()
+	{
+		void *returnCap;
+		asm volatile("cjal %0, 4" : "=C"(returnCap));
+		return returnCap;
+	}
+
+	[[cheri::interrupt_state(enabled)]] void *get_interrupts_enabled_sentry()
+	{
+		void *returnCap;
+		asm volatile("cjal %0, 4" : "=C"(returnCap));
+		return returnCap;
+	}
+
+	void test_sentries()
+	{
+		Capability interruptsEnabledSentry = {get_interrupts_enabled_sentry()};
+		debug_log("interrupts enabled sentry {}", interruptsEnabledSentry);
+		TEST(interruptsEnabledSentry.type() == 3,
+		     "Expected type 3 but got {}",
+		     interruptsEnabledSentry.type());
+
+		Capability interruptsDisabledSentry = {
+		  get_interrupts_disabled_sentry()};
+		debug_log("interrupts disabled sentry {}", interruptsDisabledSentry);
+		TEST(interruptsDisabledSentry.type() == 2,
+		     "Expected type 2 but got {}",
+		     interruptsDisabledSentry.type());
+	}
+
+	/*
+	 * Convenience function for getting the address of pointer. Should compile
+	 * down to cgetaddr.
+	 */
+	template<class T>
+	ptraddr_t address_of_pointer(T *p)
+	{
+		Capability cap{p};
+		return cap.address();
+	}
+
+	/**
+	 * Function that performs a store of capability `value` through `location`.
+	 * We call this with values we expect to fault and want to know the address
+	 * of the faulting instruction. Since it is so simple it will have no
+	 * prelude so we can use the address of the function as the address of the
+	 * store. Obviously we can't allow it to be inlined.
```

Comment:
1) I'm not sure about the names; these are not _intrinsically_ going to fault, they're just used that way ("He's not evil, he's just written that way")... maybe `perform_store` or something like that?

2) Perhaps this and `faulting_load` should be in assembler rather than relying on the (sensible, probably not going to change, but still not contractual) behavior of the compiler?